pkg/doc/jwt/jwt.go

```diff
@@ -209,24 +180,90 @@ func mapJWSToJWT(jws *jose.JSONWebSignature) (*JSONWebToken, error) {
 	}
 
 	return &JSONWebToken{
-		Headers:   headers,
-		Payload:   claims,
-		signature: jws.Signature(),
+		Headers: headers,
+		Payload: claims,
+		jws:     jws,
 	}, nil
 }
 
-// New creates new JSON Web Token based on input claims.
-func New(claims interface{}) (*JSONWebToken, error) {
-	m, err := toMap(claims)
+// NewSigned creates new signed JSON Web Token based on input claims.
+func NewSigned(claims interface{}, headers jose.Headers, signer jose.Signer) (*JSONWebToken, error) {
+	return newSigned(claims, headers, signer)
+}
+
+// NewUnsecured creates new unsecured JSON Web Token based on input claims.
+func NewUnsecured(claims interface{}, headers jose.Headers) (*JSONWebToken, error) {
+	return newSigned(claims, headers, &unsecuredJWTSigner{})
+}
+
+func newSigned(claims interface{}, headers jose.Headers, signer jose.Signer) (*JSONWebToken, error) {
+	payloadMap, err := toMap(claims)
 	if err != nil {
 		return nil, fmt.Errorf("unmarshallable claims: %w", err)
 	}
 
+	payloadBytes, err := json.Marshal(payloadMap)
+	if err != nil {
+		return nil, fmt.Errorf("marshal JWT claims: %w", err)
+	}
+
+	// JWS compact serialization uses only protected headers (https://tools.ietf.org/html/rfc7515#section-3.1).
+	jws, err := jose.NewJWS(headers, nil, payloadBytes, signer)
+	if err != nil {
+		return nil, fmt.Errorf("create JWS: %w", err)
+	}
+
 	return &JSONWebToken{
-		Payload: m,
+		Headers: jws.ProtectedHeaders,
+		Payload: payloadMap,
+		jws:     jws,
 	}, nil
 }
 
+// IsJWS checks if JWT is a JWS of valid structure.
+func IsJWS(s string) bool {
+	parts := strings.Split(s, ".")
+
+	isValidJSON := func(s string) bool {
+		b, err := base64.RawURLEncoding.DecodeString(s)
+		if err != nil {
+			return false
+		}
+
+		var j map[string]interface{}
+		err = json.Unmarshal(b, &j)
+
+		return err == nil
+	}
+
+	return len(parts) == 3 &&
+		isValidJSON(parts[0]) &&
+		isValidJSON(parts[1]) &&
+		parts[2] != ""
+}
+
+// IsJWTUnsecured checks if JWT is an unsecured JWT of valid structure.
+func IsJWTUnsecured(s string) bool {
+	parts := strings.Split(s, ".")
+
+	isValidJSON := func(s string) bool {
```

Comment:
this variable can be moved out and used by both `IsJWTUnsercured()` and `IsJWS()`

Reply:
Ah, right! I did this once but probably rewrote this change. Fixed.